src/wmtk/PolygonMesh.cpp

```diff
@@ -0,0 +1,239 @@
+        h_tuple.m_hash);
+}
+
+Tuple PolygonMesh::opp_halfedge(const Tuple& h_tuple) const
+{
+    long h = h_tuple.m_global_cid;
+    long edge_hid = h % 2;
+    return Tuple(
+        h_tuple.m_local_vid,
+        h_tuple.m_local_eid,
+        h_tuple.m_local_fid,
+        (edge_hid == 0) ? h + 1 : h - 1,
+        h_tuple.m_hash);
+}
+
+bool PolygonMesh::is_connectivity_valid() const
+{
+    // TODO
+    assert(false);
+    return true;
+}
+
+Tuple PolygonMesh::tuple_from_id(const PrimitiveType type, const long gid) const
+{
+    // TODO Need to get hashes
+
+    switch (type) {
+    case PrimitiveType::Vertex: {
+        ConstAccessor<long> vh_accessor = create_const_accessor<long>(m_vh_handle);
+        auto h = vh_accessor.index_access().scalar_attribute(gid);
+        return Tuple(0, -1, -1, h, 0);
+    }
+    case PrimitiveType::Edge: {
+        long h = 2 * gid;
+        return Tuple(0, -1, -1, h, 0);
+    }
+    case PrimitiveType::Face: {
+        ConstAccessor<long> fh_accessor = create_const_accessor<long>(m_fh_handle);
+        auto h = fh_accessor.index_access().scalar_attribute(gid);
+        return Tuple(0, -1, -1, h, 0);
+    }
+    case PrimitiveType::Tetrahedron:
+    default: throw std::runtime_error("Invalid primitive type"); break;
+    }
+}
+
+bool PolygonMesh::is_ccw(const Tuple& tuple) const
+{
+    return (tuple.m_local_vid == 0);
+}
+
+bool PolygonMesh::is_hole_face(const Tuple& f_tuple) const
+{
+    ConstAccessor<bool> f_is_hole_accessor = create_const_accessor<bool>(m_f_is_hole_handle);
+    auto f_is_hole = f_is_hole_accessor.scalar_attribute(f_tuple);
+    return f_is_hole;
+}
+
+bool PolygonMesh::is_boundary(const Tuple& tuple) const
+{
+    return is_boundary_edge(tuple);
+}
+
+bool PolygonMesh::is_boundary_vertex(const Tuple& tuple) const
+{
+    Tuple v_tuple = tuple;
+    do {
+        if (is_boundary(v_tuple)) {
+            return true;
+        }
+        v_tuple = switch_edge(switch_face(v_tuple));
+    } while (v_tuple != tuple);
+
+    return false;
+}
+
+bool PolygonMesh::is_boundary_edge(const Tuple& tuple) const
+{
+    return (is_hole_face(tuple)) || (is_hole_face(opp_halfedge(tuple)));
+}
+
+bool PolygonMesh::is_valid(const Tuple& tuple, ConstAccessor<long>& hash_accessor) const
+{
+    // TODO
+    assert(false);
+    return true;
+}
+
+void PolygonMesh::initialize(
+    Eigen::Ref<const VectorXl> next,
+    Eigen::Ref<const VectorXl> boundary_faces)
+{
+    // TODO
+    assert(false);
+}
+
+void PolygonMesh::initialize(Eigen::Ref<const RowVectors3l> F)
+{
+    // TODO
+    assert(false);
+}
+
+long PolygonMesh::id(const Tuple& tuple, PrimitiveType type) const
+{
+    switch (type) {
+    case PrimitiveType::Vertex: {
+        long h = tuple.m_global_cid;
+        long edge_hid = h % 2;
+        long edge_vid = (edge_hid + tuple.m_local_vid) % 2;
+
+        ConstAccessor<long> ev_accessor = create_const_accessor<long>(m_ev_handle);
+        auto ev = ev_accessor.vector_attribute(tuple);
```

Comment:
so do both halfedges store both of their vertex indices? ideally they would each store one?

Reply:
same with faces. i might just not be remembering your datastructure

Reply:
No, there are two faces and two vertices stored per edge (so one each per halfedge). Each halfedge should store just one vertex, either at its base or at its tip. The choice is arbitrary, but needs to be consistent. This code block is assuming the stored vertex is at the base of the halfedge (although I should probably change it to the tip to be consistent with the writeup).

The computation of the edge_vid here makes it so that swapping the local_vid between 0 and 1 also swaps the endpoint returned by `id`.